fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/JoinSkewAddSalt.java

```diff
@@ -0,0 +1,212 @@
+import org.apache.doris.nereids.trees.expressions.SlotReference;
+import org.apache.doris.nereids.trees.expressions.WhenClause;
+import org.apache.doris.nereids.trees.expressions.functions.Function;
+import org.apache.doris.nereids.trees.expressions.functions.generator.ExplodeNumbers;
+import org.apache.doris.nereids.trees.expressions.functions.scalar.If;
+import org.apache.doris.nereids.trees.expressions.functions.scalar.Random;
+import org.apache.doris.nereids.trees.expressions.literal.BigIntLiteral;
+import org.apache.doris.nereids.trees.expressions.literal.IntegerLiteral;
+import org.apache.doris.nereids.trees.expressions.literal.NullLiteral;
+import org.apache.doris.nereids.trees.plans.DistributeType;
+import org.apache.doris.nereids.trees.plans.JoinType;
+import org.apache.doris.nereids.trees.plans.Plan;
+import org.apache.doris.nereids.trees.plans.algebra.SetOperation.Qualifier;
+import org.apache.doris.nereids.trees.plans.logical.LogicalGenerate;
+import org.apache.doris.nereids.trees.plans.logical.LogicalJoin;
+import org.apache.doris.nereids.trees.plans.logical.LogicalProject;
+import org.apache.doris.nereids.trees.plans.logical.LogicalUnion;
+import org.apache.doris.nereids.types.DataType;
+import org.apache.doris.nereids.types.IntegerType;
+import org.apache.doris.nereids.types.TinyIntType;
+import org.apache.doris.nereids.util.TypeCoercionUtils;
+
+import com.google.common.collect.ImmutableList;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * LogicalJoin(type:inner, t1.a=t2.a, hint:skew(t1.a(1,2)))
+ *   +--LogicalOlapScan(t1)
+ *   +--LogicalOlapScan(t2)
+ * ->
+ * LogicalJoin (type:inner, t1.a=t2.a and r1=r2)
+ *   |--LogicalProject (t1.a, CASE WHEN t1.a IS NULL THEN random(0, 999)
+ *          WHEN t1.a IN (1, 2) THEN random(0, 999) ELSE 0 END AS r1))
+ *   | +--LogicalOlapScan(t1)
+ *   +--LogicalProject (projections: t2.a, if(explodeNumber IS NULL, 0, explodeNumber) as r2)
+ *     +--LogicalJoin (type=right_outer_join, t2.a=skewValue)
+ *       |--LogicalGenerate(generators=[explode_numbers(1000)], generatorOutput=[explodeNumber])
+ *       |  +--LogicalUnion(outputs=[skewValue], constantExprsList(1,2))
+ *       +--LogicalOlapScan(t2)
+ * */
+public class JoinSkewAddSalt extends OneRewriteRuleFactory {
+    private static final String RANDOM_COLUMN_NAME_LEFT = "r1";
+    private static final String RANDOM_COLUMN_NAME_RIGHT = "r2";
+    private static final String SKEW_VALUE_COLUMN_NAME = "skewValue";
+    private static final String EXPLODE_NUMBER_COLUMN_NAME = "explodeColumn";
+
+    @Override
+    public Rule build() {
+        return logicalJoin().thenApply(JoinSkewAddSalt::transform).toRule(RuleType.JOIN_SKEW_ADD_SALT);
+    }
+
+    private static Plan transform(MatchingContext<LogicalJoin<Plan, Plan>> ctx) {
+        LogicalJoin<Plan, Plan> join = ctx.root;
+        int factor = ctx.connectContext.getStatementContext().getConnectContext()
```

Comment:
In my understanding, this "factor" is the instance num of cluster. for example, there are 3*(16/2)=24 instances in a cluster with 3 BEs(16 core), the factor should be 24